File: spsc-lite-ruby/AdvancedProcessTreeBuilder.rb
require 'BasicProcessTreeBuilder'
require 'HE'
require 'MSG'

module ProcessTreeBuilder
  class AdvancedBuilder < BasicBuilder
    include SLL
    include ProcessTree
    include HE
    include MSG
    def initialize(drivingEngine, exp)
      @drivingEngine = drivingEngine
      @tree = Tree.new(exp)
      @nameGen = drivingEngine.nameGen
      @msgBuilder = MSGBuilder.new(@nameGen)
    end

    def abstract(alpha, exp, subst)
      bindings = subst.to_a.sort
      letExp = Let.new(exp, bindings)
      @tree.replaceSubtree(alpha, letExp)
    end

    def split(beta)
      exp = beta.exp
      args = exp.args
      names1 = @nameGen.freshNameList(args.length)
      args1 = names1.map{|x| Var.new(x)}
      letExp = Let(beta.e.cloneFunctor(args1), names1.zip(args))
      @tree.replaceSubtree(beta, letExp)
    end

    def generalizeAlphaOrSplit(beta, alpha)
      gen = @msgBuilder.build(alpha.exp, beta.exp)
      if gen.exp.isVar()
        split(beta)
      else
        abstract(alpha, gen.exp, gen.substA)
      end
    end

    def findEmbeddedAncestor(beta)
      beta.ancestors.find do |alpha|
        alpha.exp.isFGCall() && embeddedIn(alpha.exp, beta.exp)
      end
    end

    def buildStep(beta)
      # This method overrides the one in the basic version of
      # the process tree builder.
      alpha = beta.findMoreGeneralAncestor()
      if alpha
        loopBack(beta, alpha)
      else
        alpha = findEmbeddedAncestor(beta)
        if alpha
          generalizeAlphaOrSplit(beta, alpha)
        else
          expandNode(beta)
        end
      end
    end

    def AdvancedBuilder.build(nameGen, k, prog, exp)
      drivingEngine = DrivingEngine.new(nameGen, prog)
      builder = AdvancedBuilder.new(drivingEngine, exp)
      builder.buildProcessTree(k)
      builder.tree
    end
  end
end

File: spsc-lite-python/src/spsc/advanced_process_tree_builder.py
'''
Created on Aug 26, 2009

@author: Sergei Romanenko
'''

### Advanced Supercompiler with homeomorphic imbedding and generalization  

from basic_process_tree_builder import *
from he import embeddedIn
from msg import MSGBuilder

class AdvancedProcessTreeBuilder(BasicProcessTreeBuilder):

    def __init__(self, drivingEngine, exp):
        self.drivingEngine = drivingEngine
        self.tree = ProcessTree(exp)
        self.nameGen = drivingEngine.nameGen
        self.msgBuilder = MSGBuilder(self.nameGen)

    def abstract(self, alpha, exp, subst):
        bindings = list(subst.items())
        bindings.sort()
        letExp = Let(exp, bindings)
        self.tree.replaceSubtree(alpha, letExp)

    def split(self, beta):
        exp = beta.exp
        args = exp.args
        names1 = self.nameGen.freshNameList(len(args))
        args1 = [Var(x) for x in names1]
        letExp = Let(beta.e.cloneFunctor(args1), zip(names1, args))
        self.tree.replaceSubtree(beta, letExp)

    def generalizeAlphaOrSplit(self, beta, alpha):
        gen = self.msgBuilder.build(alpha.exp, beta.exp)
        if gen.exp.isVar():
            self.split(beta)
        else:
            self.abstract(alpha, gen.exp, gen.substA)

    def findEmbeddedAncestor(self, beta):
        for alpha in beta.ancestors():
            if alpha.exp.isFGCall() and embeddedIn(alpha.exp, beta.exp):
                return alpha
        return None

    def buildStep(self, beta):
        """
        This method overrides the method in the basic version of
        the process tree builder.
        """
        alpha = beta.findMoreGeneralAncestor()
        if alpha:
            self.loopBack(beta, alpha)
        else:
            alpha = self.findEmbeddedAncestor(beta)
            if alpha:
                self.generalizeAlphaOrSplit(beta, alpha)
            else:
                self.expandNode(beta)

def buildAdvancedProcessTree(nameGen, k, prog, exp):
    drivingEngine = DrivingEngine(nameGen, prog)
    builder = AdvancedProcessTreeBuilder(drivingEngine, exp)
    builder.buildProcessTree(k)
    return builder.tree
